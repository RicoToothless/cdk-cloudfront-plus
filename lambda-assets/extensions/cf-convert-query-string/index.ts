import { URLSearchParams } from 'url'
export async function lambdaHandler(event: any) {
    /**
     * Whitelisted request arguments that will be converted into additional headers
     * `NEEDED_KEYS` is the placeholder for `args` that will be bundled and injected by esbuild.
     * */
    const neededKeys = NEEDED_KEYS

    console.log(`${JSON.stringify(event, null, '\t')}`);

    let request = event.Records[0].cf.request;
    let querystring = request.querystring;

    let beforeHeaders = request.headers;
    let beforeQueryString = querystring;

    // Parse request querystring to get dictionary/json
    let params = new URLSearchParams(querystring);

    console.log(`Before processing, the content of headers:${JSON.stringify(beforeHeaders, null, '\t')}`);
    console.log(`Before processing, the content of querystring:${JSON.stringify(beforeQueryString, null, '\t')}`);
    if (params != null) {
        // Add headers according to a query string
        const capitalizeHeaderOrNot = true
        for (let key of neededKeys) {
            _add_header(key, params, request, capitalizeHeaderOrNot);
        }
    }
    let afterHeader = request.headers;
    let afterQueryString = request.querystring;
    console.log(`After processing, the content of headers:${JSON.stringify(afterHeader, null, '\t')}`);
    console.log(`After processing, the content of querystring:${JSON.stringify(afterQueryString, null, '\t')}`);

    console.log(JSON.stringify(request, null, '\t'));
    return request;
}


function capitalizeFirstLetterForEachTearm(term: string, delimiter: string = '-'): string {
    let components = term.split(delimiter);
    components.forEach((item, index) => components[index] = item.charAt(0).toUpperCase() + item.slice(1));
    let result = components.join(delimiter);
    return result;
}

function _add_header(headerName: string, params: any, request: any, capitalizeOrNot: boolean = false): void {
    const properHeader = (capitalizeOrNot == true) ? capitalizeFirstLetterForEachTearm('x-'.concat(headerName)) : headerName;
    request.headers['x-'.concat(headerName)] = [
        { "key": properHeader, "value": (params.get(headerName) === null) ? '' : params.get(headerName) }]
    params.delete(headerName);
}
