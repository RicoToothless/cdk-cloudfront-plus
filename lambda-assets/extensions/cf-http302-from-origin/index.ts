import { String } from 'aws-sdk/clients/cloudsearch';
import axios, { AxiosResponse } from 'axios';

export interface HttpResponse {
  readonly status: number;
  readonly statusDescription: string;
  readonly headers: { [key: string]: {[key: string]: String}[] },
  readonly body: string,
}

export async function HttpGet(url: string): Promise<AxiosResponse> {
  const client = axios.create();
  const response: AxiosResponse<any> = await client.get(url)
  return response
}

export async function handler(event: any) {
  const response = event.Records[0].cf.response;
  console.log('response: %j', response)
  if (response.status === '302' && response.headers['location'].length != 0) {
    const locationUrl = response.headers['location'][0].value;
    const newResponse = await HttpGet(locationUrl)
    const responseObject: HttpResponse = {
      headers: {
        'cache-control': [{
          key: 'Cache-Control',
          value: `max-age=0`,
        }],
      },
      status: newResponse.status,
      statusDescription: newResponse.statusText,
      body: newResponse.data,
    }
    return responseObject
  } else {
    return response;
  }
};


function cacheControl(responseObject: HttpResponse,  maxAge: string) {
  return Object.assign(responseObject, {
    headers: {
      'cache-control': [{
        key: 'Cache-Control',
        value: `max-age=${maxAge}`,
      }],
    }
  })
}
