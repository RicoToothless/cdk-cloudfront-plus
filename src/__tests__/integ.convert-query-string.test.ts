import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import * as cf from '@aws-cdk/aws-cloudfront';
import { HttpOrigin } from '@aws-cdk/aws-cloudfront-origins';
import * as cdk from '@aws-cdk/core';
import * as extensions from '../extensions';

test('minimal usage', () => {
  // GIVEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'demo-stack');

  // WHEN
  // create a cloudfront distribution with an extension (L@E)
  const convertQueryStringProsp: extensions.ConvertQueryStringProps = { args: ['language', 'name'] };
  const convertQueryString = new extensions.ConvertQueryString(stack, 'ConvertQueryStringDemo', convertQueryStringProsp);

  // A CloudFront distribution
  const cloudFrontDistribution = new cf.Distribution(stack, 'CloudFrontDistribution', {
    defaultBehavior: {
      origin: new HttpOrigin('postman-echo.com', {
        httpPort: 80,
        originPath: '/get',
        originSslProtocols: [cf.OriginSslPolicy.TLS_V1],
        keepaliveTimeout: cdk.Duration.seconds(10),
        protocolPolicy: cf.OriginProtocolPolicy.HTTP_ONLY,
        readTimeout: cdk.Duration.seconds(10),
      }),
      edgeLambdas: [convertQueryString],
      cachePolicy: new cf.CachePolicy(stack, 'DefaultCachePolicy', {
        cachePolicyName: 'ConvertQueryString-Cache-Policy',
        queryStringBehavior: cf.CacheQueryStringBehavior.all(),
      }),
      originRequestPolicy: new cf.OriginRequestPolicy(stack, 'RequestPolicy', {
        originRequestPolicyName: 'ConvertQueryString-Request-Policy',
        queryStringBehavior: cf.OriginRequestQueryStringBehavior.all(),
        headerBehavior: cf.OriginRequestHeaderBehavior.all(),
        comment: 'just for demonstration.',
      }),
    },
    comment: 'The CloudFront distribution based on the custom origin',
    priceClass: cf.PriceClass.PRICE_CLASS_200,
  });
  new cdk.CfnOutput(stack, 'DistributionDomainName', {
    value: cloudFrontDistribution.distributionDomainName,
  });

  // THEN
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();

  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      Comment: 'The CloudFront distribution based on the custom origin',
      DefaultCacheBehavior: {
        LambdaFunctionAssociations: [
          {
            EventType: 'origin-request',
            LambdaFunctionARN: {
              Ref: 'ConvertQueryStringFuncCurrentVersion4FB275862a9b84221e5ba9190684389f5c63a7be',
            },
          },
        ],
        ViewerProtocolPolicy: 'allow-all',
      },
      Origins: [
        {
          CustomOriginConfig: {
            HTTPPort: 80,
            OriginKeepaliveTimeout: 10,
            OriginProtocolPolicy: 'http-only',
            OriginReadTimeout: 10,
            OriginSSLProtocols: [
              'TLSv1',
            ],
          },
          DomainName: 'postman-echo.com',
          Id: 'demostackCloudFrontDistributionOrigin15405BC3B',
          OriginPath: '/get',
        },
      ],
      PriceClass: 'PriceClass_200',
    },
  },
  );
});
