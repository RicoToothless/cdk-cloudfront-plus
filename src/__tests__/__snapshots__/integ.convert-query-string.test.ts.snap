// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal usage 1`] = `
Object {
  "Outputs": Object {
    "DistributionDomainName": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "CloudFrontDistributionBA64CE3A",
          "DomainName",
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters725288a44466702e58112c016b46e32a22979f1c9721ea8cbbe647fc6f1defd8ArtifactHashDE06FA77": Object {
      "Description": "Artifact hash for asset \\"725288a44466702e58112c016b46e32a22979f1c9721ea8cbbe647fc6f1defd8\\"",
      "Type": "String",
    },
    "AssetParameters725288a44466702e58112c016b46e32a22979f1c9721ea8cbbe647fc6f1defd8S3Bucket6F316730": Object {
      "Description": "S3 bucket for asset \\"725288a44466702e58112c016b46e32a22979f1c9721ea8cbbe647fc6f1defd8\\"",
      "Type": "String",
    },
    "AssetParameters725288a44466702e58112c016b46e32a22979f1c9721ea8cbbe647fc6f1defd8S3VersionKey6A35822B": Object {
      "Description": "S3 key for asset version \\"725288a44466702e58112c016b46e32a22979f1c9721ea8cbbe647fc6f1defd8\\"",
      "Type": "String",
    },
    "AssetParametersdf679057014ef50dd1ae2f20bdb7f377a871dffb3e7b572b61076f8d30db037fArtifactHash92BBD73C": Object {
      "Description": "Artifact hash for asset \\"df679057014ef50dd1ae2f20bdb7f377a871dffb3e7b572b61076f8d30db037f\\"",
      "Type": "String",
    },
    "AssetParametersdf679057014ef50dd1ae2f20bdb7f377a871dffb3e7b572b61076f8d30db037fS3Bucket19A117E6": Object {
      "Description": "S3 bucket for asset \\"df679057014ef50dd1ae2f20bdb7f377a871dffb3e7b572b61076f8d30db037f\\"",
      "Type": "String",
    },
    "AssetParametersdf679057014ef50dd1ae2f20bdb7f377a871dffb3e7b572b61076f8d30db037fS3VersionKey5047FF25": Object {
      "Description": "S3 key for asset version \\"df679057014ef50dd1ae2f20bdb7f377a871dffb3e7b572b61076f8d30db037f\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CloudFrontDistributionBA64CE3A": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Comment": "The CloudFront distribution based on the custom origin",
          "DefaultCacheBehavior": Object {
            "CachePolicyId": Object {
              "Ref": "DefaultCachePolicyDDFA5BDC",
            },
            "Compress": true,
            "LambdaFunctionAssociations": Array [
              Object {
                "EventType": "origin-request",
                "IncludeBody": false,
                "LambdaFunctionARN": Object {
                  "Ref": "ConvertQueryStringFuncCurrentVersion4FB2758612c4fe8a1fa2539cd7ce54282ff3f624",
                },
              },
            ],
            "OriginRequestPolicyId": Object {
              "Ref": "RequestPolicyD134CC98",
            },
            "TargetOriginId": "demostackCloudFrontDistributionOrigin15405BC3B",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "CustomOriginConfig": Object {
                "HTTPPort": 80,
                "OriginKeepaliveTimeout": 10,
                "OriginProtocolPolicy": "http-only",
                "OriginReadTimeout": 10,
                "OriginSSLProtocols": Array [
                  "TLSv1",
                ],
              },
              "DomainName": "postman-echo.com",
              "Id": "demostackCloudFrontDistributionOrigin15405BC3B",
              "OriginPath": "/get",
            },
          ],
          "PriceClass": "PriceClass_200",
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "ConvertQueryStringDemoNestedStackConvertQueryStringDemoNestedStackResource0583644D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParameters725288a44466702e58112c016b46e32a22979f1c9721ea8cbbe647fc6f1defd8S3Bucket6F316730",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters725288a44466702e58112c016b46e32a22979f1c9721ea8cbbe647fc6f1defd8S3VersionKey6A35822B",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters725288a44466702e58112c016b46e32a22979f1c9721ea8cbbe647fc6f1defd8S3VersionKey6A35822B",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "ConvertQueryStringFuncCurrentVersion4FB2758612c4fe8a1fa2539cd7ce54282ff3f624": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "ConvertQueryStringFuncFA34B149",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "ConvertQueryStringFuncFA34B149": Object {
      "DependsOn": Array [
        "ConvertQueryStringFuncServiceRoleC2C97826",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersdf679057014ef50dd1ae2f20bdb7f377a871dffb3e7b572b61076f8d30db037fS3Bucket19A117E6",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdf679057014ef50dd1ae2f20bdb7f377a871dffb3e7b572b61076f8d30db037fS3VersionKey5047FF25",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdf679057014ef50dd1ae2f20bdb7f377a871dffb3e7b572b61076f8d30db037fS3VersionKey5047FF25",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.lambdaHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ConvertQueryStringFuncServiceRoleC2C97826",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ConvertQueryStringFuncServiceRoleC2C97826": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "edgelambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DefaultCachePolicyDDFA5BDC": Object {
      "Properties": Object {
        "CachePolicyConfig": Object {
          "DefaultTTL": 86400,
          "MaxTTL": 31536000,
          "MinTTL": 0,
          "Name": "ConvertQueryString-Cache-Policy",
          "ParametersInCacheKeyAndForwardedToOrigin": Object {
            "CookiesConfig": Object {
              "CookieBehavior": "none",
            },
            "EnableAcceptEncodingBrotli": false,
            "EnableAcceptEncodingGzip": false,
            "HeadersConfig": Object {
              "HeaderBehavior": "none",
            },
            "QueryStringsConfig": Object {
              "QueryStringBehavior": "all",
            },
          },
        },
      },
      "Type": "AWS::CloudFront::CachePolicy",
    },
    "RequestPolicyD134CC98": Object {
      "Properties": Object {
        "OriginRequestPolicyConfig": Object {
          "Comment": "just for demonstration.",
          "CookiesConfig": Object {
            "CookieBehavior": "none",
          },
          "HeadersConfig": Object {
            "HeaderBehavior": "allViewer",
          },
          "Name": "ConvertQueryString-Request-Policy",
          "QueryStringsConfig": Object {
            "QueryStringBehavior": "all",
          },
        },
      },
      "Type": "AWS::CloudFront::OriginRequestPolicy",
    },
  },
}
`;
